lib/Interpreter/IncrementalCUDADeviceCompiler.h

```diff
@@ -0,0 +1,204 @@
+//--------------------------------------------------------------------*- C++ -*-
+// CLING - the C++ LLVM-based InterpreterG :)
+// author:  Simeon Ehrig <simeonehrig@web.de>
+//
+// This file is dual-licensed: you can choose to license it under the University
+// of Illinois Open Source License or the GNU Lesser General Public License. See
+// LICENSE.TXT for details.
+//------------------------------------------------------------------------------
+
+#ifndef CLING_INCREMENTAL_CUDA_DEVICE_JIT_H
+#define CLING_INCREMENTAL_CUDA_DEVICE_JIT_H
+
+#include "clang/Frontend/CodeGenOptions.h"
+#include "llvm/ADT/SmallVector.h"
+
+#include <string>
+#include <vector>
+
+namespace cling{
+    class InvocationOptions;
+    class Transaction;
+}
+
+namespace clang {
+    class CompilerInstance;
+    class HeaderSearchOptions;
+    class LangOptions;
+}
+
+namespace llvm {
+    class StringRef;
+}
+
+namespace cling {
+
+  ///\brief The class is responsible for generating CUDA device code in
+  /// cuda fatbinary form during the runtime. It works with
+  /// llvm::sys::ExecuteAndWait.
+  ///
+  class IncrementalCUDADeviceCompiler {
+
+    static constexpr unsigned CxxStdCompiledWith() {
+      // The value of __cplusplus in GCC < 5.0 (e.g. 4.9.3) when
+      // either -std=c++1y or -std=c++14 is specified is 201300L, which fails
+      // the test for C++14 or more (201402L) as previously specified.
+      // I would claim that the check should be relaxed to:
+#if __cplusplus > 201402L
+      return 17;
+#elif __cplusplus > 201103L || (defined(LLVM_ON_WIN32) && _MSC_VER >= 1900)
+      return 14;
+#elif __cplusplus >= 201103L
+      return 11;
+#else
+#error "Unknown __cplusplus version"
+#endif
+    }
+
+    ///\brief Contains the arguments for the cling nvptx and the nvidia
+    /// fatbinary tool. The arguments are static and will set at the constructor
+    /// of IncrementalCUDADeviceCompiler.
+    struct CUDACompilerArgs {
+      std::string cppStdVersion = "-std=c++" + std::to_string(CxxStdCompiledWith());
```

Comment:
Instead "invalid", I would use an empty string. So the clang can use the default c++ standard. In general, it is more a theoretically problem, because the standard of the cling will be used and it is c++ 11 by default.

Reply:
Let's make sure that we see an error, please. Whether returning `"invalid"` or `assert()`ing - anything works for me. But we should make sure that we notice that this switch needs to be updated once C++20 arrives.